SHA512/224 and SHA512/256

paulidale opened the pull request:
Support added for these two digests, available only via the EVP interface.

- [x] tests are added or updated

bernd-edlinger commented:
documentation?

dot-asm commented on crypto/include/internal/sha.h:
```diff
@@ -0,0 +1,31 @@
+
+# ifdef  __cplusplus
+extern "C" {
+# endif
```
What's the point of extern "C" in *internal* header. Users won't have opportunity to compile it with C++ compiler...

paulidale replied:
Good point.  I'll fix.

dot-asm commented on test/recipes/30-test_evp_data/evpdigest.txt:
```diff
@@ -167,6 +167,62 @@ Input = "a"
 
+Digest = SHA512-224
+Input = 
+Output = 6ed0dd02806fa89e25de060c19d3ac86cabb87d6a0ddd05c333b84f4
```
Reference to source would be more than appropriate. I mean where the test vectors come from.

paulidale replied:
The vectors are from NIST, I'll add references tomorrow.

dot-asm commented on crypto/sha/sha512.c:
```diff
@@ -59,6 +60,42 @@
 # define SHA512_BLOCK_CAN_MANAGE_UNALIGNED_DATA
 #endif
 
+int sha512_224_init(SHA512_CTX *c)
```
I don't recall that lower-case would be reflection of symbol visibility... I mean I don't think there is a rule that says that unexported symbols **has to be*** lower case...

paulidale replied:
I was in a bit of a quandary over this and other nuances.

The move seems to be towards not exporting lower level functions, thus I changed the the names to lowercase which seems to (mostly) be the convention.  I also didn't export the evp_sha512_224() function, instead only allowing access via the string representation.

I'm happy to change either or to export everything but we'll want consensus first.

richsalz replied:
I think the EVP function should be available; the others are.  I think all-lowercase-names for internal functions is common, even if not a formal standard of ours so I would stick with that.

paulidale commented:
@bernd-edlinger I'm not sure where or what to document.  There are no exported functions here and the documentation pages deal with these.  Suggestions are welcome.

I'm happy to add something to the CHANGES file if that's deemed appropriate.  I feel that it would be but not strongly.

richsalz commented:
If you make the EVP API public, then it could go in one of the EVP_sha manpages.

paulidale commented:
Fixed things as per comments.

Added test sources, removed the C++ wrapper, made the EVP_sha512_224 and EVP_sha512_256 functions public and added documentation.

paulidale commented:
Comments, suggestions or approval anyone?

paulidale commented:
@dot-asm ?

paulidale commented:
Rebased to trunk -- the only conflict being in the `libcrypto.num` file.

dot-asm commented:
> @dot-asm ?

I don't quite follow what is expected from me. My approval doesn't make difference in another committer's request, so in cases when review outcome is insignificant as a rule I'm reluctant to add one. As in this one, just nits...

paulidale commented:
Since you'd reviewed already, I figured it would be polite to get your confirmation as well.

paulidale commented:
Merged, thanks.
